### ISSUE TITLE
Obfuscation for wireguard

### ISSUE THREAD

**Issue by DikozImpact** [throneproj/Throne#219]
The Russian government has learned how to block Wireguard. It requires obfuscation to work now
Please add obfuscation settings for wireguard configs as in AmneziaWG (https://github.com/amnezia-vpn/amneziawg-go?ysclid=m3bnacvhcj541033059) or in Hiddify (https://github.com/hiddify/hiddify-app)

At the moment NekoRay can only work with regular wireguard, and it does not support obfuscation parameters.

**Comment by Mahdi-zarei** [comment]
Will try to add it, but for now I'm too busy...

**Comment by imagecreation1** [comment]
+1. Here's the link for your comfort: https://hiddify.com/app/How-to-use-WARP-on-Hiddify-App/.
It would be great if we could use these bypass filtering modes and noise packets to fool DPI.

**Comment by Mahdi-zarei** [comment]
I am a bit confused now, amnezia uses its own layers over wireguard, which makes it incompatible with the mainstream wireguard implementation. Hiddify uses other methods, but I have heard mixed ( towards negative ) news of its success. What implementation does exactly work better? If anyone has data on this I would like to test it before using it in nekoray.

**Comment by DikozImpact** [comment]
> I am a bit confused now, amnezia uses its own layers over wireguard, which makes it incompatible with the mainstream wireguard implementation. Hiddify uses other methods, but I have heard mixed ( towards negative ) news of its success. What implementation does exactly work better? If anyone has data on this I would like to test it before using it in nekoray.

What kind of data do you need exactly?

**Comment by Mahdi-zarei** [comment]
> What kind of data do you need exactly?

Which implementation is widely used and in what countries it has been successful.

**Comment by DikozImpact** [comment]
> Which implementation is widely used and in what countries it has been successful.

I can only speak for Russia. AmneziaWG is more popular than hiddify, besides hiddify really has problems with bugs. Probably obfuscation from Hiddify will be easier to implement, because it is built on sing-box core, but in my opinion Amnezia is a better option

**Comment by imagecreation1** [comment]
As far as I know, AmneziaWG's obfuscation isn't compatible with vanilla Wireguard configs, which makes it kind of redundant. Correct me if I'm wrong. 
I think it will be better for compatibility and availability to implement Hiddify's obfuscation.

**Comment by KDBal** [comment]
> Correct me if I'm wrong.

@imagecreation1 you are somewhat wrong.
Generally speaking, AWG is not comptatible with regular WG, but it can work in "compatibility mode".
For that you need to set AWG parameters to:
S1=0
S2=0
H1=1
H2=2
H3=3
H4=4
then it can successfully connect to a regular WG server (although, with weaker obfuscation).

**Comment by Verity-Freedom** [comment]
@Mahdi-zarei

> > What kind of data do you need exactly?
> 
> Which implementation is widely used and in what countries it has been successful.

Well Amnezia is a Russian company that created WG protocol modification called amnezia WG which is WIDELY used in Russia to bypass censorship to already blocked Wireguard, for example WARP on wireguard is blocked in Russia, but on AWG isn't. The main tools which support AWG is an amnezia VPN (their VPN for many protocols) and amnezia WG (old wireguard client modification). WG and AWG are compatible and may be transitioned between each other painlessly. I don't know why you heard mixed results about AWG because it may be already used by millions, mainly to bypass Discord censorship. I see it all inside because I'm Russian.

**Comment by Verity-Freedom** [comment]
> AmneziaWG's obfuscation isn't compatible with vanilla Wireguard configs

It's not true because the obfuscation is built in layers of additional parameters. Once you removed those parameters, you use the ordinary wireguard, and vice versa. There is a program called amnezia wg tools on github which is used for WG -> AWG conversion.

<pre>[Interface]
PrivateKey = EOtJky3/ESEGuIA5C27oVtBNugeIO8LeteKCFTUyDnU=
Address = 172.16.0.2, 2606:4700:110:8013:1301:994a:7153:88c6
DNS = 1.1.1.1, 2606:4700:4700::1111, 1.0.0.1, 2606:4700:4700::1001
MTU = 1280

[Peer]
PublicKey = bmXOC+F1FxEMF9dyiK2H5/1SUtzH0JuVo51h2wPfgyo=
AllowedIPs = 0.0.0.0/0, ::/0
Endpoint = engage.cloudflareclient.com:2408</pre>

This is wireguard to WARP.

<pre>[Interface]
PrivateKey = EOtJky3/ESEGuIA5C27oVtBNugeIO8LeteKCFTUyDnU=
S1 = 0
S2 = 0
Jc = 120
Jmin = 23
Jmax = 911
H1 = 1
H2 = 2
H3 = 3
H4 = 4
Address = 172.16.0.2, 2606:4700:110:8013:1301:994a:7153:88c6
DNS = 1.1.1.1, 2606:4700:4700::1111, 1.0.0.1, 2606:4700:4700::1001
MTU = 1280

[Peer]
PublicKey = bmXOC+F1FxEMF9dyiK2H5/1SUtzH0JuVo51h2wPfgyo=
AllowedIPs = 0.0.0.0/0, ::/0
Endpoint = engage.cloudflareclient.com:2408</pre>

This is AWG to WARP.

**Comment by Verity-Freedom** [comment]
@Mahdi-zarei I believe this will be useful.

**Comment by Verity-Freedom** [comment]
> then it can successfully connect to a regular WG server

I think any AWG may connect to WG servers without any problems, but you can check it. I ask to not use Hiddify because Hiddify is, in general, a bad program. Amnezia WG client is compatible both to wireguard and amnezia WG, you can use it to check.
https://github.com/amnezia-vpn/amneziawg-windows-client/releases/tag/1.0.0

**Comment by Verity-Freedom** [comment]
I don't know anything about Hiddify obfuscation, maybe it's not a bad one too. I just don't like Hiddify program in general, I think it lacks some important capabilities.

**Comment by Verity-Freedom** [comment]
> towards negative ) news of its success

I just didn't hear anything about hiddify wireguard obfuscation, but feels like it is a blatant copy of AWG. You can try to implement both, of course, but AWG is 100% much older (at least April 2024) and much more popular.

**Comment by Verity-Freedom** [comment]
@Mahdi-zarei and thats WG - AWG conversion tool.
https://github.com/teslaproduuction/Amnezia-WireGuard-Tools/releases/tag/v.1.0.0

**Comment by DikozImpact** [comment]
@Verity-Freedom Еб**ь ты тут насрал. Да ещё и читать не умеешь. 
> I don't know why you heard mixed results about AWG 
Он говорил про хидди, а не амнезию. Зачем-то капсом орёшь по миллионы пользователей, хотя это вообще ничего не значит. Она популярна просто потому что самая простая и распиаренная, но полезного функционала там 0. У нас тот же planetvpn популярен, хотя это говно то ещё, с кучей рекламмы, которое поскачивали мамочки вбив в поисковик "бесплатный впн". Кучу воды налил бесполезной. Зачем-то обвиняешь хидди в воровстве, хотя он существует ещё с 2022 года. Зачем-то про амнезиювпн написал, хотя это такая же кривая поделка как и хидди. 

@Mahdi-zarei Don't listen carefully to the person above who doesn't understand the subject. I like Amnesia better too, of course, but he's talking nonsense and bullshit in many ways.

**Comment by DikozImpact** [comment]
> @Verity-Freedom Еб**ь ты тут насрал. Да ещё и читать не умеешь.
> 
> > I don't know why you heard mixed results about AWG
> 
> Он говорил про хидди, а не амнезию. Зачем-то капсом орёшь по миллионы пользователей, хотя это вообще ничего не значит. Она популярна просто потому что самая простая и распиаренная, но полезного функционала там 0. У нас тот же planetvpn популярен, хотя это говно то ещё, с кучей рекламы, которое поскачивали мамочки вбив в поисковик "бесплатный впн". Кучу воды налил бесполезной. Зачем-то обвиняешь хидди в воровстве, хотя он существует ещё с 2022 года. Зачем-то про амнезиювпн написал, хотя это такая же кривая поделка как и хидди. В общем лучше не пиши ничего, если в теме разбираешься на уровне ребёнка из тиктока.
> 
> @Mahdi-zarei Don't listen carefully to the person above who doesn't understand the subject. I like Amnezia better too, of course, but he's talking nonsense and bullshit in many ways. Even if you add obfuscation from Hiddify I'll be glad because something is better than nothing

**Comment by Verity-Freedom** [comment]
> Еб**ь ты тут насрал

I'm MALWARE chat mod (the person from you forked your only repository), you did not say anything new, you did not say anything where I was wrong and why. I don't understand what do you mean at all, and you can't just curse at me in Russian with asterisks without ANY reason to behave this ruse, this is a complete nonsense, Mahdi obviously doesn't know Russian. We were talking not about programs hiddify and Amnezia VPN, but about protocols. People here said many wrong things about AWG protocol, especially about incompatibility between WG and AWG, in fact they're compatible with both servers and conversion in one to another WITH WINDOWS NOTEBOOK REDACTION and AWG tools program, to which I provided the link above. I checked in Webarchive the page about Hiddify wireguard obfuscation, and the most early mention of it is the august 2024, that is why I said the AWG protocol is most likely older than Hiddify obfuscation, I don't 100% sure is this true or not.

https://web.archive.org/web/20240824092933/https://hiddify.com/app/How-to-use-WARP-on-Hiddify-App/

**Comment by Verity-Freedom** [comment]
> хотя это такая же кривая поделка как и хидди

Amnezia VPN AT LEAST has split tunneling unlike Hiddify, but again, we were talking not about programs, but about protocols, I'm a nekoray user. I said that I think hiddify stole the idea of wireguard obfuscation from Amnezia and I stand by this.

**Comment by Verity-Freedom** [comment]
> Зачем-то обвиняешь хидди в воровстве, хотя он существует ещё с 2022 года

Ye you most likely didn't read at all what I wrote, kid.

I said that hiddify most likely stole the idea of AWG protocol to create their own poor realization of Wireguard obfuscation, the fact that hiddify was created in 2022 has nothing to do with their wireguard obfuscation methods. You just said above that Amnezia is a better option, and I wrote exactly the same thing, why to argue with me about it? I don't know any single person in Russia which are saying that hiddify wireguard obfuscation is better than AWG protocol while being a MALWARE chat mod, a person that is popularised AWG protocol in Russia. Learn how to read and don't argue with people if in the end you just agree with them in key points.

**Comment by DikozImpact** [comment]
_Facepalm_

**Comment by KDBal** [comment]
> hiddify stole the idea of wireguard obfuscation from Amnezia

@Verity-Freedom The idea of sending fake packages for avoiding WG detection was known long before both AWG and Hiddify implementations. You can even use curl for that.

This is irrelevant to the topic.
Nobody is suggesting forking hiddify GUI (no matter how good or bad it is). The logic of obfuscation is contained inside their sing-box fork and, therefore, can be trivially ported to nekobox (which already uses sing-box itself).
AWG, on the other hand, does come only with a separate binary and would require either adding it as an additional core (which would prevent from using an existing wireguard implementation inside sing-box), or implementing the logic behind it yourself (which is somewhat tricky, since there is not even a decent documentation on the matter).

Also, the only things that AWG suggests with respect to original WG protocol are fake packets (JC, Jmin, Jmax) and hiddify has a more versatile configuration, therefore more advanced. On the other hand, then we are willing to break WG compatibility, AWG becomes a better solution.

**Comment by Verity-Freedom** [comment]
> On the other hand, then we are willing to break WG compatibility, AWG becomes a better solution.

That is why I said that AWG is fully compatible to WG unlike people above, when they said they are not compatible, it is simply not true. Maybe hiddify obfuscation will be easier to implement but again as a large TG chat mode (there are 1868 people), I don't know any single person that is using hiddify instead of AWG obfuscation. The fact that there is more parameters in hiddify doesn't mean it's better. And I don't know anything about the ideas to obfuscate wireguard before amnezia team, which created AWG circa April 2024. Mahdi said that hiddify obfuscation isn't working so good based on the information that they have.

**Comment by DikozImpact** [comment]
@KDBal I don't think you should continue this discussion, it's not worth it to be honest.

**Comment by KDBal** [comment]
@Mahdi-zarei If You are not willing to spend a lot of time on the issue, I would suggest taking a look at [ByeDPI](https://github.com/hufrea/byedpi) project. It also provides an obfuscation for wireguard (although, the weakest of mentioned here) and is trivial to add (as an additional core for example). It also provides methods of obfuscating regular TLS traffic, which is somewhat useful.

**Comment by Verity-Freedom** [comment]
> @Mahdi-zarei If You are not willing to spend a lot of time on the issue, I would suggest taking a look at [ByeDPI](https://github.com/hufrea/byedpi) project. It also provides an obfuscation for wireguard (although, the weakest of mentioned here) and is trivial to add (as an additional core for example). It also provides methods of obfuscating regular TLS traffic, which is somewhat useful.

Well I believe that AWG and Hiddify obfuscation are much similar, both adding noise/trash packets, but it is still important to understand that 1) Hiddify stole this idea 2) We don't know much about hiddify efficiency, unlike AWG, which proved to be good 3) Hiddify uses more parameters that can ruin WG compatibility.

![image](https://github.com/user-attachments/assets/64d8df0d-1e2e-45ab-8d22-68f092ad10bf)

![image](https://github.com/user-attachments/assets/a62aa663-6d3b-41df-9c15-b519e6c824ec)

**Comment by Verity-Freedom** [comment]
S1 = 0
S2 = 0
Jc = X
Jmin = X
Jmax = X
H1 = 1
H2 = 2
H3 = 3
H4 = 4

That's really all of AWG, where X is a fairly random number, other numbers are unchanged. You just edit those in notebook in a conf file, you can delete or add them at any moment, and they are easily generated by AWG tools. This is 100% compatibility, and WE ALL USE ONLY AWG OBFUSCATION IN RUSSIA, NOT HIDDIFY ATTEMPT TO COPY IT.
https://github.com/teslaproduuction/Amnezia-WireGuard-Tools/releases/tag/v.1.0.0

**Comment by Verity-Freedom** [comment]
@DikozImpact that's a simple question for you as for Russian, does hiddify WARP obfuscation works for you, or are you using only AWG, like all of us? That would be really nice to know.

**Comment by Verity-Freedom** [comment]
> Он говорил про хидди, а не амнезию

Yes I misread at that point, I'm sorry, I corrected myself. He talked about mixed and negative opinions about hiddify obfuscation, not AWG, but this actually only proves my point further, that maybe hiddify obfuscation is not that good, is not used anywhere, and you yourself constantly says that Amnezia is better, just like me. If you use hiddify wireguard obfuscation, then say about your opinion, I asked you to do this one message above. I'm using AWG and I know that it is working fairly great not only for WARP, but also for Proton, for FineVPN (there is a full backwards wireguard compatibility), which is why I'm saying it is so great.

**Comment by KDBal** [comment]
@Verity-Freedom
Please, could You send less? We do understand the logic behind AWG.
Mahdi-zarei is a decent developer and can figure things out on his own.
Every time you write something a notification is sent to all subscribed to this issue and that is distracting.

Hiddify implementation is relatively new, but it is used in Iran.
It is also used in Russia by hiddify users then they connect to WARP. I can confirm that it works.

The thing worth mentioning that the moment DPI systems start ignoring fake packages all this configs would stop working. In that case implementing AWG is more worth, since it would still remain useful (although only for AWG servers) while other implementations would just become garbage.

**Comment by Verity-Freedom** [comment]
> It is also used in Russia by hiddify users then they connect to WARP. I can confirm that it works.

That's great, that's what I asked for. Well the best thing to do is to implement support for both, probably, but it obviously will take some time to implement even one.

**Comment by Verity-Freedom** [comment]
> who "stole" what open-source protocol

Well at least hiddify implementation is working, I just don't think they made AWG better than it was at the start. You can't just rename VLESS into VLECC and say that it is now a better VLESS version. I didn't feel so good when I found out today that hiddify at some point reskinned AWG which I already used for a couple of months.

**Comment by Verity-Freedom** [comment]
Turns out different people are working on this since at least 2022 so...
https://github.com/infinet/xt_wgobfs

Well now it's obvious it's a very useful thing with many different implementations.

**Comment by Mahdi-zarei** [comment]
Has anyone had experiences with https://github.com/database64128/swgp-go ? My tests show that Iran's firewall seemingly identifies wireguard from its handshake packet's patterns, and while sending junk packets may help, it is only a temporary solution, as they can extend the dpi to investigate further packets. swgp fully conceals the handshake and thus can be a much better choice, its only downside is that it acts as an intermediary layer and requires another instance of swgp on the server side and thus is incompatible with the vanilla wireguard. But I assume for those hosting their own VPN servers this is not an issue and swgp provides a more reliable way of hiding wireguard.

**Comment by Botan626** [comment]
Whu noise for WG was not added? Sad.

**Comment by ndiruhniu** [comment]
I think this issue can be closed since amneziawg has been added in beta 6.

**Comment by bropines** [comment]
> Has anyone had experiences with [database64128/swgp-go](https://github.com/database64128/swgp-go?rgh-link-date=2025-04-23T05%3A22%3A34.000Z) ? My tests show that Iran's firewall seemingly identifies wireguard from its handshake packet's patterns, and while sending junk packets may help, it is only a temporary solution, as they can extend the dpi to investigate further packets. swgp fully conceals the handshake and thus can be a much better choice, its only downside is that it acts as an intermediary layer and requires another instance of swgp on the server side and thus is incompatible with the vanilla wireguard. But I assume for those hosting their own VPN servers this is not an issue and swgp provides a more reliable way of hiding wireguard.

https://github.com/amnezia-vpn/amneziawg-go/releases/tag/v0.2.14-beta-awg-1.5-1

Will there be support for AWG-1.5?

**Comment by Verity-Freedom** [comment]
> > Has anyone had experiences with [database64128/swgp-go](https://github.com/database64128/swgp-go?rgh-link-date=2025-04-23T05%3A22%3A34.000Z) ? My tests show that Iran's firewall seemingly identifies wireguard from its handshake packet's patterns, and while sending junk packets may help, it is only a temporary solution, as they can extend the dpi to investigate further packets. swgp fully conceals the handshake and thus can be a much better choice, its only downside is that it acts as an intermediary layer and requires another instance of swgp on the server side and thus is incompatible with the vanilla wireguard. But I assume for those hosting their own VPN servers this is not an issue and swgp provides a more reliable way of hiding wireguard.
> 
> https://github.com/amnezia-vpn/amneziawg-go/releases/tag/v0.2.14-beta-awg-1.5-1
> 
> Will there be support for AWG-1.5?

Better to create a new issue probably. Amnezia still didn't release full specs for their proto.
